### PR TITLE
refactor(ext/http): rewrite hyper integration and fix bug

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

File: ext/http/lib.rs [hunk 1]
```diff
@@ -73,96 +83,255 @@ pub fn init() -> Extension {
...
+        None => self.closed().map_ok(|_| None).await,
+      }
+    }
+    .try_or_cancel(&self.cancel_handle)
+    .await
+  }
+
+  /// A future that completes when this HTTP connection is closed or errors.
+  async fn closed(&self) -> Result<(), AnyError> {
+    self.closed_fut.clone().map_err(AnyError::from).await
```
[the remark this answers is not in the document]
Comment (AaronO) [reply ext/http/lib.rs:168]: I agree, I'm also planning a refactor on error mapping

**Comment by piscisaureus** [comment]
> Have you done benchmarks yet?


<details>
  <summary>[CI run 1](https://github.com/denoland/deno/runs/4015450253)</summary> 

``` json
{
  "req_per_sec": {
    "hyper": 62084,
    "node_tcp": 49991,
    "deno_http": 29683,
    "core_http_json_ops": 59498,
    "node_http": 20366,
    "deno_http_native": 29477,
    "deno_tcp": 55574
  },
  "max_latency": {
    "deno_tcp": 1.93,
    "deno_http_native": 1.49,
    "node_http": 1.85,
    "deno_http": 1.34,
    "hyper": 1.25,
    "node_tcp": 0.93,
    "core_http_json_ops": 1.05
  }
}
```
</detail>

**Comment by ry** [comment]
# Performance Analysis

Using:  `./target/release/deno run --allow-net cli/bench/deno_http_native.js` and  `wrk --latency -d 20s http://127.0.0.1:4500`

Summary: throughput degraded by 9%, latency seems to be marginally better

## Before
```
Running 20s test @ http://127.0.0.1:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    98.23us   93.73us   7.42ms   99.25%
    Req/Sec    48.71k     1.74k   50.13k    99.00%
  Latency Distribution
     50%   93.00us
     75%  111.00us
     90%  133.00us
     99%  183.00us
  1948484 requests in 20.10s, 161.67MB read
Requests/sec:  96941.69
Transfer/sec:      8.04MB

Running 20s test @ http://127.0.0.1:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    97.74us   45.58us   5.43ms   90.10%
    Req/Sec    48.39k     1.75k   53.19k    98.26%
  Latency Distribution
     50%   94.00us
     75%  111.00us
     90%  134.00us
     99%  185.00us
  1935425 requests in 20.10s, 160.58MB read
Requests/sec:  96292.27
Transfer/sec:      7.99MB
```

## After

```
Running 20s test @ http://127.0.0.1:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   102.64us   60.99us   5.27ms   98.18%
    Req/Sec    46.16k     2.08k   47.89k    97.51%
  Latency Distribution
     50%   93.00us
     75%  112.00us
     90%  131.00us
     99%  182.00us
  1846511 requests in 20.10s, 153.20MB read
Requests/sec:  91865.10
Transfer/sec:      7.62MB

Running 20s test @ http://127.0.0.1:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   103.10us   32.60us   4.22ms   87.10%
    Req/Sec    45.47k     1.61k   47.91k    93.78%
  Latency Distribution
     50%   98.00us
     75%  115.00us
     90%  133.00us
     99%  173.00us
  1818720 requests in 20.10s, 150.90MB read
Requests/sec:  90482.88
Transfer/sec:      7.51MB

```

File: cli/tests/unit/http_test.ts [hunk 2]
```diff
@@ -999,3 +1001,39 @@ unitTest(
...
+      for (let i = 0; i < 10; i++) {
+        const response = await fetch(`http://${hostname}:${port}/${i}`);
+        const body = await response.text();
+        assertStrictEquals(body, `Response #${i}`);
+      }
+    }
+
+    await Promise.all([server(), delay(100).then(client)]);
+  },
+);
```
Comment (ry) [inline cli/tests/unit/http_test.ts:1039]: Nice test. I have verified this fails on main.

**Comment by piscisaureus** [comment]
Add test for HTTP2 requests where the second one gets a response before the first one.

File: runtime/web_worker.rs [hunk 3]
```diff
@@ -587,6 +585,5 @@ pub fn run_web_worker(
     debug!("Worker thread shuts down {}", &name);
     result
   };
-
-  rt.block_on(fut)
+  run_basic(fut)
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply runtime/web_worker.rs:588]: Necessary, but in a separate commit.

File: ext/http/01_http.js [hunk 4]
```diff
@@ -79,26 +83,26 @@
...
 
       const [
-        requestRid,
-        responseSenderRid,
+        streamRid,
         method,
         headersList,
         url,
       ] = nextRequest;
+      SetPrototypeAdd(this.managedResources, streamRid);
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply ext/http/01_http.js:97]: The `managedResources` set is associated with an HttpConn instance. It contains multiple `HttpStreamResource`s.

File: ext/http/lib.rs [hunk 5]
```diff
@@ -73,96 +83,255 @@ pub fn init() -> Extension {
...
+        Either::Right((_, mut conn_fut)) => {
+          conn_fut.as_mut().graceful_shutdown();
+          conn_fut.await
+        }
+      };
+      filter_enotconn(result).map_err(Arc::from)
+    };
+    let (task_fut, closed_fut) = task_fut.remote_handle();
+    let closed_fut = closed_fut.shared();
+    spawn_local(task_fut);
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply ext/http/lib.rs:128]: > Very fancy, but apparently working.

What would you suggest that's less fancy?

File: ext/http/lib.rs [hunk 6]
```diff
@@ -73,96 +83,255 @@ pub fn init() -> Extension {
...
+    }
+  }
+
+  // Accepts a new incoming HTTP request.
+  async fn accept(
+    self: &Rc<Self>,
+  ) -> Result<Option<HttpStreamResource>, AnyError> {
+    let fut = async {
+      let (request_tx, request_rx) = oneshot::channel();
+      let (response_tx, response_rx) = oneshot::channel();
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply ext/http/lib.rs:145]: I don't think so.

**Review by ry — APPROVED** [review]
LGTM

* Combining the resources is a big clean up
* The rust side of the code is much easier to reason about.
* This also fixes a bug

However this patch also degrades throughput performance by about 4% on hello-world server. I've done another perf analysis below. We think the benefits of this refactor outweigh the slowness. We hope this serves as good foundation for further improvements in `ext/http`.

# Performance Analysis 2

Using:  `./target/release/deno run --allow-net cli/bench/deno_http_native.js` and  `wrk --latency -d 20s http://127.0.0.1:4500`

Summary: throughput degraded by 4%, latency seems to be marginally better.


## Deno v1.15.3
```
Running 20s test @ http://127.0.0.1:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    99.53us   99.42us   7.86ms   99.30%
    Req/Sec    48.21k     1.69k   50.66k    99.00%
  Latency Distribution
     50%   94.00us
     75%  112.00us
     90%  135.00us
     99%  186.00us
  1928035 requests in 20.10s, 159.97MB read
Requests/sec:  95922.25
Transfer/sec:      7.96MB

Running 20s test @ http://127.0.0.1:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    98.50us   37.24us   3.58ms   82.77%
    Req/Sec    47.92k     1.61k   49.61k    99.00%
  Latency Distribution
     50%   95.00us
     75%  113.00us
     90%  136.00us
     99%  187.00us
  1916882 requests in 20.10s, 159.04MB read
Requests/sec:  95370.21
Transfer/sec:      7.91MB
```

## This Patch

```
Running 20s test @ http://127.0.0.1:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   101.77us   32.60us   3.81ms   83.89%
    Req/Sec    46.00k     1.40k   47.69k    98.01%
  Latency Distribution
     50%  100.00us
     75%  115.00us
     90%  131.00us
     99%  167.00us
  1840014 requests in 20.10s, 152.67MB read
Requests/sec:  91543.34
Transfer/sec:      7.60MB

Running 20s test @ http://127.0.0.1:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   101.91us   29.00us   3.19ms   79.30%
    Req/Sec    45.90k     1.40k   47.43k    98.26%
  Latency Distribution
     50%  101.00us
     75%  115.00us
     90%  132.00us
     99%  167.00us
  1835615 requests in 20.10s, 152.30MB read
Requests/sec:  91323.73
Transfer/sec:      7.58MB
```

# Node v17.0.1 baseline

```
Running 20s test @ http://127.0.0.1:4544
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   100.67us  171.98us  12.41ms   99.67%
    Req/Sec    50.30k     2.86k   53.15k    98.01%
  Latency Distribution
     50%   89.00us
     75%  115.00us
     90%  150.00us
     99%  209.00us
  2011687 requests in 20.10s, 257.08MB read
Requests/sec: 100084.42
Transfer/sec:     12.79MB

Running 20s test @ http://127.0.0.1:4544
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    98.57us  114.68us   9.48ms   99.17%
    Req/Sec    50.33k     2.81k   52.86k    97.76%
  Latency Distribution
     50%   89.00us
     75%  115.00us
     90%  150.00us
     99%  208.00us
  2012718 requests in 20.10s, 257.21MB read
Requests/sec: 100134.06
Transfer/sec:     12.80MB
```

**Pull request merged**